### PR TITLE
Add ProgressManager on CC

### PR DESCRIPTION
**Pull request by arukuka**
# Summary

- Changed so that a progress can be returned to the client
- Related to https://github.com/acompany-develop/QuickMPC-libClient-py/pull/13

# Purpose

This patch makes that client will be able to get job progress

# Contents

- Add `ProgressManager` class to manager progress
- Update gRPC protobuf files
- Replace previous progress logging

# Testing Methods Performed

- Run `demo/intergration_demo/progress_demo.py`
  - with https://github.com/acompany-develop/QuickMPC-libClient-py/pull/13

### CODE REVIEW

File: src/ComputationContainer/Job/ProgressManager.hpp [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        {
+            this->id = id;
+            return *this;
+        }
+        Builder& setDesc(const std::string& desc)
+        {
+            this->description = desc;
+            return *this;
+        }
+        Builder& setObserver(const std::shared_ptr<Observer> observer)
```
Comment (arukuka) [inline src/ComputationContainer/Job/ProgressManager.hpp:33]:
```suggestion
        Builder& setObserver(const std::shared_ptr<Observer>& observer)
```

File: src/ComputationContainer/Job/ProgressManager.hpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    /**
+     * instantiate Progress from Builder.
+     * bacause all parameters is needed,
+     * this constructor throw `std::bad_optional_access`
+     * if even one parameter is not set.
+     */
+    Progress(const Builder& builder);
+
+    /**
+     * represent progress by [0.0, 1.0].
```
Comment (arukuka) [inline src/ComputationContainer/Job/ProgressManager.hpp:57]:
```suggestion
     * represent progress by [0.0, 100.0].
```

File: src/Proto/common_types/common_types.proto [hunk 3]
```diff
@@ -33,3 +33,17 @@ enum PredictMethod {
     PREDICT_METHOD_DECISION_TREE = 3;
     PREDICT_METHOD_SID3_TREE = 4;
 }
+
+message ProcedureProgress {
+    uint64 id = 1;                // unique id in ProgressManager context
+    string description = 2;       // describe specific process in job
+    float progress = 3;           // describe progress by [0.0, 1.0]
```
Comment (arukuka) [inline src/Proto/common_types/common_types.proto:40]:
```suggestion
    float progress = 3;           // describe progress by [0.0, 100.0]
```

**Review by mdonaka — COMMENTED** [review]
Thank you for the big change!
A few corrections please.

**Review by KotaTakahashi9320 — APPROVED** [review]

File: src/ComputationContainer/Client/ComputationToDb/ValueTable.cpp [hunk 4]
```diff
@@ -150,80 +165,87 @@ std::vector<std::pair<int, int>> intersectionSortedValueIndex(
...
-            else if (comp[comp_it])
-            {
-                ++now_its[i].first;
-            }
-            else
+
+            // [0th less, 0th greater, 1st less, 1st greater ....]
+            auto comp = qmpc::Share::allLess(comp_l, comp_r);
+            auto comp_it = 0;
+            auto progress = 100.0;
```
[the remark this answers is not in the document]
Comment (arukuka) [reply src/ComputationContainer/Client/ComputationToDb/ValueTable.cpp]: @mdonaka Thank you for your review.  I applied the patch.

File: src/ComputationContainer/Client/ComputationToDb/ValueTable.cpp [hunk 5]
```diff
@@ -150,80 +165,87 @@ std::vector<std::pair<int, int>> intersectionSortedValueIndex(
...
+                }
+                else
+                {
+                    ++now_its[i].second;
+                }
+
+                auto max_progress =
+                    std::max(now_its[i].first - block_it[i], now_its[i].second - upper[i]) * 100.0
+                    / block_size;
+                progress = std::min(progress, max_progress);
```
[the remark this answers is not in the document]
Comment (arukuka) [reply src/ComputationContainer/Client/ComputationToDb/ValueTable.cpp]: @mdonaka Thank you for your review.  I applied the patch.

**Pull request merged**